File: src/transform/ecef.ts
/**
 * ECEF (Earth-Centered Earth-Fixed) Coordinate Calculations
 *
 * Implements WGS84 ellipsoid calculations and East-North-Up to ECEF transforms.
 * This is equivalent to Cesium's Transforms.eastNorthUpToFixedFrame().
 */

import { createLogger } from '../utils/logger.js';
import type { WGS84Position } from '../types.js';
import type { Matrix4, Vector3 } from './matrix.js';

const logger = createLogger('ecef');

// WGS84 Ellipsoid Parameters
const WGS84 = {
  // Semi-major axis (equatorial radius) in meters
  a: 6378137.0,

  // Semi-minor axis (polar radius) in meters
  b: 6356752.314245,

  // Flattening
  f: 1 / 298.257223563,

  // First eccentricity squared
  e2: 0.00669437999014,

  // Second eccentricity squared
  ep2: 0.00673949674228,
};

/**
 * Convert degrees to radians
 */
function toRadians(degrees: number): number {
  return degrees * (Math.PI / 180);
}

/**
 * Convert WGS84 geodetic coordinates (lon, lat, height) to ECEF Cartesian coordinates
 */
export function wgs84ToEcef(position: WGS84Position): Vector3 {
  const lon = toRadians(position.longitude);
  const lat = toRadians(position.latitude);
  const h = position.height;

  const sinLon = Math.sin(lon);
  const cosLon = Math.cos(lon);
  const sinLat = Math.sin(lat);
  const cosLat = Math.cos(lat);

  // Radius of curvature in the prime vertical
  const N = WGS84.a / Math.sqrt(1 - WGS84.e2 * sinLat * sinLat);

  // ECEF coordinates
  const x = (N + h) * cosLat * cosLon;
  const y = (N + h) * cosLat * sinLon;
  const z = (N * (1 - WGS84.e2) + h) * sinLat;

  return [x, y, z];
}

/**
 * Convert ECEF Cartesian coordinates to WGS84 geodetic coordinates
 */
export function ecefToWgs84(ecef: Vector3): WGS84Position {
  const [x, y, z] = ecef;

  // Calculate longitude
  const longitude = Math.atan2(y, x) * (180 / Math.PI);

  // Iterative calculation for latitude (Bowring's method)
  const p = Math.sqrt(x * x + y * y);
  let lat = Math.atan2(z, p * (1 - WGS84.e2));

  for (let i = 0; i < 10; i++) {
    const sinLat = Math.sin(lat);
    const N = WGS84.a / Math.sqrt(1 - WGS84.e2 * sinLat * sinLat);
    const newLat = Math.atan2(z + WGS84.e2 * N * sinLat, p);
    if (Math.abs(newLat - lat) < 1e-12) break;
    lat = newLat;
  }

  const latitude = lat * (180 / Math.PI);

  // Calculate height
  const sinLat = Math.sin(lat);
  const cosLat = Math.cos(lat);
  const N = WGS84.a / Math.sqrt(1 - WGS84.e2 * sinLat * sinLat);
  const height = p / cosLat - N;

  return { longitude, latitude, height };
}

/**
 * Calculate the surface normal (up direction) at a WGS84 position
 * This points away from the ellipsoid surface (geodetic normal)
 */
export function surfaceNormal(position: WGS84Position): Vector3 {
  const lon = toRadians(position.longitude);
  const lat = toRadians(position.latitude);

  const cosLon = Math.cos(lon);
  const sinLon = Math.sin(lon);
  const cosLat = Math.cos(lat);
  const sinLat = Math.sin(lat);

  // Geodetic surface normal
  return [cosLat * cosLon, cosLat * sinLon, sinLat];
}

/**
 * Compute the East-North-Up to ECEF transformation matrix
 *
 * This creates a 4x4 matrix that transforms coordinates from a local
 * East-North-Up (ENU) frame centered at the given position to ECEF coordinates.
 *
 * The ENU frame has:
 * - X axis pointing East
 * - Y axis pointing North
 * - Z axis pointing Up (away from ellipsoid surface)
 *
 * This is equivalent to Cesium's Transforms.eastNorthUpToFixedFrame()
 */
export function eastNorthUpToEcef(position: WGS84Position): Matrix4 {
  const lon = toRadians(position.longitude);
  const lat = toRadians(position.latitude);

  const cosLon = Math.cos(lon);
  const sinLon = Math.sin(lon);
  const cosLat = Math.cos(lat);
  const sinLat = Math.sin(lat);

  // East vector (tangent to parallel, pointing east)
  const east: Vector3 = [-sinLon, cosLon, 0];

  // North vector (tangent to meridian, pointing north)
  const north: Vector3 = [-sinLat * cosLon, -sinLat * sinLon, cosLat];

  // Up vector (geodetic surface normal, pointing away from Earth)
  const up: Vector3 = [cosLat * cosLon, cosLat * sinLon, sinLat];

  // Origin in ECEF
  const origin = wgs84ToEcef(position);

  // Build 4x4 transformation matrix (column-major order)
  // | east.x  north.x  up.x  origin.x |
  // | east.y  north.y  up.y  origin.y |
  // | east.z  north.z  up.z  origin.z |
  // |   0       0       0       1     |
  const matrix: Matrix4 = [
    east[0], east[1], east[2], 0,
    north[0], north[1], north[2], 0,
    up[0], up[1], up[2], 0,
    origin[0], origin[1], origin[2], 1,
  ];

  logger.debug(
    {
      position,
      origin,
      east,
      north,
      up,
    },
    'Computed ENU-to-ECEF transform'
  );

  return matrix;
}

/**
 * Compute the ECEF to East-North-Up transformation matrix
 * (inverse of eastNorthUpToEcef)
 */
export function ecefToEastNorthUp(position: WGS84Position): Matrix4 {
  const lon = toRadians(position.longitude);
  const lat = toRadians(position.latitude);

  const cosLon = Math.cos(lon);
  const sinLon = Math.sin(lon);
  const cosLat = Math.cos(lat);
  const sinLat = Math.sin(lat);

  // These are the rows of the inverse rotation matrix (transpose of rotation)
  const east: Vector3 = [-sinLon, cosLon, 0];
  const north: Vector3 = [-sinLat * cosLon, -sinLat * sinLon, cosLat];
  const up: Vector3 = [cosLat * cosLon, cosLat * sinLon, sinLat];

  const origin = wgs84ToEcef(position);

  // Compute the translation in ENU frame
  const tx = -(east[0] * origin[0] + east[1] * origin[1] + east[2] * origin[2]);
  const ty = -(north[0] * origin[0] + north[1] * origin[1] + north[2] * origin[2]);
  const tz = -(up[0] * origin[0] + up[1] * origin[1] + up[2] * origin[2]);

  // Build inverse matrix (column-major)
  return [
    east[0], north[0], up[0], 0,
    east[1], north[1], up[1], 0,
    east[2], north[2], up[2], 0,
    tx, ty, tz, 1,
  ];
}

/**
 * Calculate the distance between two ECEF points
 */
export function ecefDistance(a: Vector3, b: Vector3): number {
  const dx = b[0] - a[0];
  const dy = b[1] - a[1];
  const dz = b[2] - a[2];
  return Math.sqrt(dx * dx + dy * dy + dz * dz);
}

/**
 * Calculate the WGS84 ellipsoid height at a point (for reference)
 */
export function ellipsoidHeightAtLatitude(latitude: number): number {
  const lat = toRadians(latitude);
  const sinLat = Math.sin(lat);
  const cosLat = Math.cos(lat);

  // Radius at latitude (distance from center to surface)
  const r = Math.sqrt(
    (Math.pow(WGS84.a * WGS84.a * cosLat, 2) +
      Math.pow(WGS84.b * WGS84.b * sinLat, 2)) /
      (Math.pow(WGS84.a * cosLat, 2) + Math.pow(WGS84.b * sinLat, 2))
  );

  return r;
}

File: src/transform/ecef.rs
/// WGS84 semi-major axis in metres.
const WGS84_A: f64 = 6_378_137.0;
/// WGS84 flattening.
const WGS84_F: f64 = 1.0 / 298.257_223_563;
/// WGS84 first eccentricity squared: e² = 2f - f²
const WGS84_E2: f64 = 2.0 * WGS84_F - WGS84_F * WGS84_F;

/// Convert geodetic (longitude, latitude, altitude) to ECEF XYZ.
///
/// Inputs are in **degrees** and metres.  Returns `[X, Y, Z]` in metres.
pub fn geodetic_to_ecef(lon_deg: f64, lat_deg: f64, alt_m: f64) -> [f64; 3] {
    let lon = lon_deg.to_radians();
    let lat = lat_deg.to_radians();

    let sin_lat = lat.sin();
    let cos_lat = lat.cos();
    let sin_lon = lon.sin();
    let cos_lon = lon.cos();

    // Radius of curvature in the prime vertical
    let n = WGS84_A / (1.0 - WGS84_E2 * sin_lat * sin_lat).sqrt();

    let x = (n + alt_m) * cos_lat * cos_lon;
    let y = (n + alt_m) * cos_lat * sin_lon;
    let z = (n * (1.0 - WGS84_E2) + alt_m) * sin_lat;

    [x, y, z]
}

/// Build the 4×4 East-North-Up rotation matrix for a given geodetic point.
///
/// Returns a column-major `[f64; 16]` matrix suitable for `tileset.json`
/// `root.transform`.
pub fn enu_rotation_matrix(lon_deg: f64, lat_deg: f64) -> [f64; 16] {
    let lon = lon_deg.to_radians();
    let lat = lat_deg.to_radians();

    let sin_lon = lon.sin();
    let cos_lon = lon.cos();
    let sin_lat = lat.sin();
    let cos_lat = lat.cos();

    // ENU basis vectors expressed in ECEF:
    //   East  = (-sin_lon,          cos_lon,         0       )
    //   North = (-sin_lat*cos_lon, -sin_lat*sin_lon, cos_lat )
    //   Up    = ( cos_lat*cos_lon,  cos_lat*sin_lon, sin_lat )
    //
    // Column-major 4×4 (translation = 0 here; combined in build_root_transform)
    #[rustfmt::skip]
    let m = [
        // column 0 (East)
        -sin_lon,
        cos_lon,
        0.0,
        0.0,
        // column 1 (North)
        -sin_lat * cos_lon,
        -sin_lat * sin_lon,
        cos_lat,
        0.0,
        // column 2 (Up)
        cos_lat * cos_lon,
        cos_lat * sin_lon,
        sin_lat,
        0.0,
        // column 3 (translation placeholder)
        0.0,
        0.0,
        0.0,
        1.0,
    ];

    m
}

/// Combine an ECEF origin and an ENU rotation into a 4×4 root transform.
///
/// Result is a column-major matrix that places the model at `ecef_origin`
/// with local axes oriented to ENU.
pub fn build_root_transform(ecef_origin: [f64; 3], enu_matrix: [f64; 16]) -> [f64; 16] {
    let mut m = enu_matrix;
    // Set the translation column (column 3, rows 0-2)
    m[12] = ecef_origin[0];
    m[13] = ecef_origin[1];
    m[14] = ecef_origin[2];
    m
}

/// Return the 4×4 identity matrix (column-major).
pub fn identity_transform() -> [f64; 16] {
    #[rustfmt::skip]
    let m = [
        1.0, 0.0, 0.0, 0.0,
        0.0, 1.0, 0.0, 0.0,
        0.0, 0.0, 1.0, 0.0,
        0.0, 0.0, 0.0, 1.0,
    ];
    m
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn geodetic_to_ecef_equator_prime_meridian() {
        // (0°, 0°, 0m) should give (a, 0, 0)
        let ecef = geodetic_to_ecef(0.0, 0.0, 0.0);
        assert!((ecef[0] - WGS84_A).abs() < 1.0); // X ≈ 6378137
        assert!(ecef[1].abs() < 1.0);               // Y ≈ 0
        assert!(ecef[2].abs() < 1.0);               // Z ≈ 0
    }

    #[test]
    fn geodetic_to_ecef_north_pole() {
        // (0°, 90°, 0m) → X≈0, Y≈0, Z≈b (semi-minor axis)
        let ecef = geodetic_to_ecef(0.0, 90.0, 0.0);
        let b = WGS84_A * (1.0 - WGS84_F); // ≈ 6356752.3
        assert!(ecef[0].abs() < 1.0);
        assert!(ecef[1].abs() < 1.0);
        assert!((ecef[2] - b).abs() < 1.0);
    }

    #[test]
    fn geodetic_to_ecef_known_point() {
        // London (51.5074° N, 0.1278° W, 0m)
        let ecef = geodetic_to_ecef(-0.1278, 51.5074, 0.0);
        // Expected approximately: X≈3978000, Y≈-8700, Z≈4968000
        assert!((ecef[0] - 3_978_000.0).abs() < 1000.0);
        assert!((ecef[1] - (-8700.0)).abs() < 1000.0);
        assert!((ecef[2] - 4_968_000.0).abs() < 1000.0);
    }

    #[test]
    fn geodetic_to_ecef_with_altitude() {
        let ecef_ground = geodetic_to_ecef(0.0, 0.0, 0.0);
        let ecef_high = geodetic_to_ecef(0.0, 0.0, 1000.0);
        // At equator, prime meridian, altitude adds to X only
        assert!((ecef_high[0] - ecef_ground[0] - 1000.0).abs() < 1.0);
    }

    #[test]
    fn enu_matrix_at_equator_prime_meridian() {
        let m = enu_rotation_matrix(0.0, 0.0);
        // East  = (0, 1, 0)  → column 0
        assert!(m[0].abs() < 1e-10);      // -sin(0) = 0
        assert!((m[1] - 1.0).abs() < 1e-10); // cos(0) = 1
        assert!(m[2].abs() < 1e-10);

        // North = (0, 0, 1)  → column 1
        assert!(m[4].abs() < 1e-10);      // -sin(0)*cos(0) = 0
        assert!(m[5].abs() < 1e-10);      // -sin(0)*sin(0) = 0
        assert!((m[6] - 1.0).abs() < 1e-10); // cos(0) = 1

        // Up    = (1, 0, 0)  → column 2
        assert!((m[8] - 1.0).abs() < 1e-10); // cos(0)*cos(0) = 1
        assert!(m[9].abs() < 1e-10);
        assert!(m[10].abs() < 1e-10);
    }

    #[test]
    fn enu_matrix_at_north_pole() {
        let m = enu_rotation_matrix(0.0, 90.0);
        // At the north pole, Up should point along +Z (ECEF)
        // Up = (cos(90)*cos(0), cos(90)*sin(0), sin(90)) = (0, 0, 1)
        assert!(m[8].abs() < 1e-10);
        assert!(m[9].abs() < 1e-10);
        assert!((m[10] - 1.0).abs() < 1e-10);
    }

    #[test]
    fn build_root_transform_sets_translation() {
        let ecef = [100.0, 200.0, 300.0];
        let enu = enu_rotation_matrix(0.0, 0.0);
        let rt = build_root_transform(ecef, enu);
        assert!((rt[12] - 100.0).abs() < 1e-10);
        assert!((rt[13] - 200.0).abs() < 1e-10);
        assert!((rt[14] - 300.0).abs() < 1e-10);
        assert!((rt[15] - 1.0).abs() < 1e-10);
    }

    #[test]
    fn identity_transform_is_correct() {
        let m = identity_transform();
        for i in 0..4 {
            for j in 0..4 {
                let expected = if i == j { 1.0 } else { 0.0 };
                assert!((m[j * 4 + i] - expected).abs() < 1e-15);
            }
        }
    }
}
